src/cmds/restic/integration_test.go

```diff
@@ -1011,3 +1011,40 @@ func TestPrune(t *testing.T) {
 		testRunCheck(t, gopts)
 	})
 }
+
+func TestHardLink(t *testing.T) {
+	withTestEnvironment(t, func(env *testEnvironment, gopts GlobalOptions) {
+		datafile := filepath.Join("testdata", "test.hl.tar.gz")
+		fd, err := os.Open(datafile)
+		if os.IsNotExist(errors.Cause(err)) {
+			t.Skipf("unable to find data file %q, skipping", datafile)
+			return
+		}
+		OK(t, err)
+		OK(t, fd.Close())
+
+		testRunInit(t, gopts)
+
+		SetupTarTestFixture(t, env.testdata, datafile)
+		opts := BackupOptions{}
+
+		// first backup
+		testRunBackup(t, []string{env.testdata}, opts, gopts)
+		snapshotIDs := testRunList(t, "snapshots", gopts)
+		Assert(t, len(snapshotIDs) == 1,
+			"expected one snapshot, got %v", snapshotIDs)
+
+		testRunCheck(t, gopts)
+
+		// restore all backups and compare
+		for i, snapshotID := range snapshotIDs {
+			restoredir := filepath.Join(env.base, fmt.Sprintf("restore%d", i))
+			t.Logf("restoring snapshot %v to %v", snapshotID.Str(), restoredir)
+			testRunRestore(t, gopts, restoredir, snapshotIDs[0])
+			Assert(t, directoriesEqualContents(env.testdata, filepath.Join(restoredir, "testdata")),
```

Comment:
This test is not enough: It only tests that the link count for each item in the the restored directory is the same as in the archived directory. This only verifies that files which were hardlinks before, are hardlinks afterwards, but does not test whether the same files are hardlinks of each other.

Suppose we have the files `A` and `B`, and the files `a`, which is a hard link of `A` and `b`, which is a hard link of `B`: `A` <-> `a` and `B` <-> `b`.

The test will succeed for `A` <-> `a` and `B` <-> `b` and also for `A` <-> `b` and `B` <-> `a`, because only the link count is verified.

I think we need to make a list before and after backup and check that the right files are linked together.

Do you agree?